Add README for Metrics (#4528)
open-telemetry/opentelemetry-dotnet: docs/metrics/README.md

# OpenTelemetry .NET Metrics

## Best Practices

> - When emitting metrics with dimensions, ALWAYS provide the dimensions in an

Reviewer: lets also add that "Instruments must be created once and re-used throughout the application.", and link to 2 ways to do that.
1. Static in the console app example.
2. DI using the asp.net core example.

Reviewer 2: Could we change this to say, `ALWAYS provide dimensions in same order` rather than sorted order to keep it simple. As the overhead is only the first time we see a new combination.

Author: Address both of the above points.

>   ascending sorted order (order by the dimension key) for best performance.

Reviewer 3: best -> better?

```csharp

> // Sorted in an ascending order by dimension key: `color` comes before `name`

Reviewer: should we stick to the term attribute or tag, instead of dimensions?

Author: Okay. Let's only use tags.

MyFruitCounter.Add(1, new("color", "red"), new("name", "apple")); // <--- DO THIS

MyFruitCounter.Add(1, new("name", "apple"), new("color", "red")); // <--- DON'T DO THIS
```

> - When emitting metrics with more than three dimensions, use `TagList` for best
>   performance.

Reviewer 3: ```suggestion
- When emitting metrics with more than three dimensions, use `TagList` for better performance (due to the stack allocation nature which avoided GC pressure).
```

```csharp

> var tags = new TagList

Reviewer: Also hint that, SDK offers highest performance for up to 3 tags, high performance for up to 8 tags, and for tags > 8, there is heap-allocation in the hot path.

Add a tip that "if you are exceeding 8 tags for a metric, check if some of the attributes are better modelled as Resources - machine name, environment etc. Those should be modelled as Resource (and link to Resource doc).".

{
    { "DimName1", "DimValue1" },
    { "DimName2", "DimValue2" },
    { "DimName3", "DimValue3" },
    { "DimName4", "DimValue4" },
};

> // Uses a TagList as there are more than three dimensions

Reviewer: ```suggestion
// Uses a TagList as there are more than three Tags
```

counter.Add(100, tags); // <--- DO THIS

// Avoid the below mentioned approaches when there are more than three dimensions
var tag1 = new KeyValuePair<string, object>("DimName1", "DimValue1");
var tag2 = new KeyValuePair<string, object>("DimName2", "DimValue2");
var tag3 = new KeyValuePair<string, object>("DimName3", "DimValue3");
var tag4 = new KeyValuePair<string, object>("DimName4", "DimValue4");

counter.Add(100, tag1, tag2, tag3, tag4); // <--- DON'T DO THIS

var readOnlySpanOfTags = new KeyValuePair<string, object>[4] { tag1, tag2, tag3, tag4};
counter.Add(100, readOnlySpanOfTags); // <--- DON'T DO THIS
```

## Common Issues that lead to missing metrics

- The `Meter` used to create the instruments is not added to the
  `MeterProvider`. Use `AddMeter` method to enable the processing for the
  required metrics.

> - Instrument name is invalid. When naming instruments, ensure that the name you

Reviewer: it maybe worthwhile to call out couple of common cases on invalid instrument names.
https://github.com/open-telemetry/opentelemetry-dotnet/issues/4117

" " (spaces in name)
Special charactors in name ("/", very common in windows performance counters)

  choose meets the criteria defined in the
  [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#instrument-name-syntax).
- MetricPoint limit is reached. By default, the SDK limits the number of maximum
  MetricPoints (unique combination of keys and values for a given Metric stream)

>   to 2000. This limit can be configured using

Reviewer 3: minor:

```suggestion
  to `2000`. This limit can be configured using
```

  `SetMaxMetricPointsPerMetricStream` method. Refer to this
  [doc](../../docs/metrics/customizing-the-sdk/README.md#changing-maximum-metricpoints-per-metricstream)
  for more information. The SDK would not process any newer unique key-value

>   combination that it encounters, once this limit is reached.

Reviewer 3: This line will change once the SDK is able to "forget" things while using DELTA?

Author: Yes. We would have to update it when that's done.

> - MeterProvider is disposed. You need to ensure that the `MeterProvider`

Reviewer: Please link to examples of asp.net core, asp.net apps showing one possible way to achieve this recommendation.

>   instance is kept active for metrics to be collected. In a typical application,

Reviewer 3: Maybe mention the console and asp.net core examples and point folks to the best practices?

# OpenTelemetry .NET Metrics

## Best Practices

- Instruments SHOULD only be created once and reused throughout the application
  lifetime. This
  [example](../../docs/metrics/getting-started-console/Program.cs) shows how an

>   instrument is created a `static` field and then used in the application. You

Reviewer: ```suggestion
  instrument is created and stored in a `static` field and then used in the application. You
```

  could also look at this ASP .NET Core
  [example](../../examples/AspNetCore/Program.cs) which shows a more Dependency

>   Injection friendly way of doing this by extracting the `Meter` and an

Reviewer: which shows a more Dependency Injection friendly way of doing this. -> I think this much is sufficient, to avoid too much details here.

  information.

## Common issues that lead to missing metrics

- The `Meter` used to create the instruments is not added to the
  `MeterProvider`. Use `AddMeter` method to enable the processing for the
  required metrics.
- Instrument name is invalid. When naming instruments, ensure that the name you
  choose meets the criteria defined in the
  [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#instrument-name-syntax).
  A few notable characters that are not allowed in the instrument name: `/`
  (forward slash), `\` (backward slash), any space character in the name.
- MetricPoint limit is reached. By default, the SDK limits the number of maximum
  MetricPoints (unique combination of keys and values for a given Metric stream)
  to `2000`. This limit can be configured using
  `SetMaxMetricPointsPerMetricStream` method. Refer to this
  [doc](../../docs/metrics/customizing-the-sdk/README.md#changing-maximum-metricpoints-per-metricstream)
  for more information. The SDK would not process any newer unique key-value
  combination that it encounters, once this limit is reached.

> - MeterProvider is disposed. You need to ensure that the `MeterProvider`

Reviewer: ```suggestion
- MeterProvider is disposed incorrectly/earlier than desired. You need to ensure that the `MeterProvider`
```

  instance is kept active for metrics to be collected. In a typical application,
  a single MeterProvider is built at application startup, and is disposed of at
  application shutdown. For an ASP.NET Core application, use `AddOpenTelemetry`
  and `WithMetrics` methods from the `OpenTelemetry.Extensions.Hosting` package

>   to correctly setup `MeterProvider`. Here's a [sample ASP .NET Core

Reviewer 3: sample -> example
ASP .NET -> ASP.NET

  app](../../examples/AspNetCore/Program.cs) for reference. For simpler

>   applications such as Console apps, refer to this

Reviewer 3: ```suggestion
  applications such as console apps, refer to this
```

>   [example](../../docs/metrics/getting-started-console/Program.cs).

Reviewer: also show asp.net example from controib repo.
